homeassistant/helpers/script.py

```diff
@@ -121,180 +132,82 @@ class _SuspendScript(Exception):
-            return self._referenced_devices
-
-        referenced = set()
-
-        for step in self.sequence:
-            action = _determine_action(step)
-
-            if action == ACTION_CHECK_CONDITION:
-                referenced |= condition.async_extract_devices(step)
-
-            elif action == ACTION_DEVICE_AUTOMATION:
-                referenced.add(step[CONF_DEVICE_ID])
-
-        self._referenced_devices = referenced
-        return referenced
+    def _config_cache(self):
+        return self._script._config_cache  # pylint: disable=protected-access
 
     @property
-    def referenced_entities(self):
-        """Return a set of referenced entities."""
-        if self._referenced_entities is not None:
-            return self._referenced_entities
-
-        referenced = set()
-
-        for step in self.sequence:
-            action = _determine_action(step)
-
-            if action == ACTION_CALL_SERVICE:
-                data = step.get(service.CONF_SERVICE_DATA)
-                if not data:
-                    continue
-
-                entity_ids = data.get(ATTR_ENTITY_ID)
+    def name(self):
+        """Return script name."""
+        return self._script.name
 
-                if entity_ids is None:
-                    continue
-
-                if isinstance(entity_ids, str):
-                    entity_ids = [entity_ids]
-
-                for entity_id in entity_ids:
-                    referenced.add(entity_id)
-
-            elif action == ACTION_CHECK_CONDITION:
-                referenced |= condition.async_extract_entities(step)
+    @property
+    def last_action(self):
+        """Return last action."""
+        return self._last_action
 
-            elif action == ACTION_ACTIVATE_SCENE:
-                referenced.add(step[CONF_SCENE])
+    @last_action.setter
+    def last_action(self, value):
+        self._last_action = value
 
-        self._referenced_entities = referenced
-        return referenced
-
-    def run(self, variables=None, context=None):
+    @abstractmethod
+    async def async_run(self) -> None:
         """Run script."""
-        asyncio.run_coroutine_threadsafe(
-            self.async_run(variables, context), self.hass.loop
-        ).result()
-
-    async def async_run(
-        self, variables: Optional[Sequence] = None, context: Optional[Context] = None
-    ) -> None:
-        """Run script.
 
-        This method is a coroutine.
-        """
-        self.last_triggered = date_util.utcnow()
-        if self._cur == -1:
-            self._log("Running script")
-            self._cur = 0
-
-        # Unregister callback if we were in a delay or wait but turn on is
-        # called again. In that case we just continue execution.
-        self._async_remove_listener()
-
-        for cur, action in islice(enumerate(self.sequence), self._cur, None):
+    async def _async_run(self, start):
+        for self._step, self._action in islice(
```

Comment:
We should add a `self._should_stop` that is an asyncio.Event.

At the start of each loop we should check if the variable exists, stop and set the event if it is. That way the `_BlockingScriptRun` can have their `stop` method stop gracefully (after the current action) instead of cancelling the running running task. What do you think of that?

Now stop can be implemented as 

```
async def async_stop(self):
    self._should_stop = asyncio.Event()
    await self._should_stop.wait()
```

Reply:
Interesting idea. I did struggle a bit with the idea that a run could get stopped no matter what it was currently doing. And since it could be calling into who-knows-what integration...

It does make a lot of sense to have a more controlled stop.

Which makes me think again about timeouts. As I think I mentioned in a comment in the previous PR, what to do about the service call's 10 second limit???

If I understand correctly how it works, the call will return after no more than 10 seconds, however the callee isn't canceled. So now how do we handle scripts that can take longer than 10 seconds? I suppose there may have been a way for that to be true before, but given the behavior of suspending for delays & waits, it was not very likely. But now it will be very likely. I guess that's something for one of the follow up PRs to address.

In any case, I'll try to work this suggestion into the re-work using `asyncio.current_task`.

Reply:
I think that it's for a follow-up PR, but I think that we could make the timeout overridable/disable the timeout.